### PR TITLE
BLOGS-1326 Parallel Guzzle Requests

### PR DESCRIPTION
**Pull request by mattgreenham**
https://jira.dev.bbc.co.uk/browse/BLOGS-1326

### CODE REVIEW

**Comment by programmes-developers** [comment]
Can one of the admins verify this patch?

File: src/Ds/Molecule/DatePicker/DatePickerPresenter.php [hunk 1]
```diff
@@ -51,7 +51,7 @@ public function getPostsYearRange(): array
         $latestYear = (int) $this->datePicker->getLatestPostDate()->format('Y');
 
         $years = [];
-        for ($year = $oldestYear; $year <= $latestYear; $year++) {
+        for ($year = $latestYear; $year >= $oldestYear; $year--) {
```
Comment (votemike) [inline src/Ds/Molecule/DatePicker/DatePickerPresenter.php]: What was the need to reverse this?

File: src/BlogsService/Repository/PostRepository.php [hunk 2]
```diff
@@ -48,6 +48,58 @@ public function getPostByGuid(string $guid, string $blogId = ''): ?ResponseInter
         return $this->getResponse($query);
     }
 
+    public function getPostsBetweenParallel(
```
Comment (votemike) [inline src/BlogsService/Repository/PostRepository.php]: Does this replace `getPostsBetween()`? Should you delete it? Can they both be used? I'm not sure it's useful to include parallel in the method name. What do you think?

File: src/BlogsService/Service/PostService.php [hunk 3]
```diff
@@ -59,51 +59,70 @@ function () use ($guidString, $blogId) {
...
+        $cacheKey = $this->cache->keyHelper(__CLASS__, __FUNCTION__, $blog->getId(), $nowDate->getTimestamp(), $page, $perpage, $ttl, $nullTtl);
 
         return $this->cache->getOrSet(
             $cacheKey,
             $ttl,
-            function () use ($blog, $publishedDate, $page, $perpage) {
+            function () use ($blog, $nowDate, $page, $perpage) {
+
+                $ranges = [
+                    'oldestPost' => ['afterDate' => Chronos::create(1970, 1, 1), 'beforeDate' => $nowDate, 'sort' => 'asc'],
```
Comment (votemike) [inline src/BlogsService/Service/PostService.php]: No need to create the same Chronos object twice.

File: src/Controller/PostByDateController.php [hunk 4]
```diff
@@ -30,15 +29,12 @@ public function __invoke(Request $request, Blog $blog, int $year, int $month, Po
 
         $nowDateTime = Chronos::now();
 
-        /** @var Post[] $latestPosts */
-        $latestPosts = $postService->getPostsByBlog($blog, $nowDateTime, 1, 1, 'desc')->getDomainModels();
-        /** @var Post[] $oldestPosts */
-        $oldestPosts = $postService->getPostsByBlog($blog, $nowDateTime, 1, 1, 'asc')->getDomainModels();
+        $oldestAndLatestPosts = $postService->getOldestPostAndLatestPost($blog, $nowDateTime);
```
Comment (votemike) [inline src/Controller/PostByDateController.php]: Would it be better to do something like `[$oldestPost, $latestPost] = $postService->getOldestPostAndLatestPost($blog, $nowDateTime);`?

File: src/Controller/PostShowController.php [hunk 5]
```diff
@@ -39,18 +40,12 @@ public function __invoke(Blog $blog, string $guid, PostService $postService, Com
 
         $this->otherIstatsLabels = $istatsLabels;
 
-        $previousPost = $postService->getPostsBefore(
-            $blog,
-            $post->getPublishedDate()
-        );
+        $prevNextPosts = $postService->getPreviousAndNextPosts($blog, $post->getPublishedDate());
```
Comment (votemike) [inline src/Controller/PostShowController.php]: Similar to above suggestion

File: src/Ds/Molecule/DatePicker/DatePickerPresenter.php (hunk 1 above)
Comment (mattgreenham) [reply src/Ds/Molecule/DatePicker/DatePickerPresenter.php]: It was showing the years in the opposite order compared to v4

File: src/BlogsService/Repository/PostRepository.php (hunk 2 above)
Comment (mattgreenham) [reply src/BlogsService/Repository/PostRepository.php]: It doesn't replace it, as it is still used by `getPostsByBlog()` and `getPostsByMonth()`. Included parallel in the name so there was an obvious way to differentiate, open to suggestions.

File: src/Ds/Molecule/DatePicker/DatePickerPresenter.php (hunk 1 above)
Comment (votemike) [reply src/Ds/Molecule/DatePicker/DatePickerPresenter.php]: Ahh, so a bug in the first place?

Comment (mattgreenham) [reply src/Ds/Molecule/DatePicker/DatePickerPresenter.php]: Yep

File: src/BlogsService/Repository/AbstractRepository.php [hunk 6]
```diff
@@ -57,12 +57,11 @@ protected function getParallelResponses(array $queries): array
                 $results[$key] = $promise->wait();
             } catch (GuzzleException $e) {
                 if ($e instanceof ClientException && $e->getCode() == 404) {
-                    $results[$key] = null;
+                    return [];
```
Comment (mattgreenham) [inline src/BlogsService/Repository/AbstractRepository.php]: @BjornTwachtmann I _think_ this is what we want to do here?

File: src/BlogsService/Service/PostService.php [hunk 7]
```diff
@@ -80,14 +80,14 @@ function () use ($blog, $publishedDate, $page, $perpage) {
...
-                $responses = $this->repository->getPostsBetweenParallel($blog->getId(), $ranges, 0, $page, $perpage);
+                $responses = $this->repository->getPostsBetween($blog->getId(), $ranges, 0, $page, $perpage);
                 $result = [];
                 foreach ($responses as $key => $response) {
                     $post = $this->responseHandler->getIsiteResult($response);
                     $result[$key] = $post->getDomainModels()[0] ?? null;
                 }
 
-                return $result;
+                return [$result['previousPost'], $result['nextPost']];
```
Comment (votemike) [inline src/BlogsService/Service/PostService.php]: Are we *sure* previousPost and nextPost will be set?

File: src/BlogsService/Service/PostService.php [hunk 8]
```diff
@@ -108,21 +108,22 @@ public function getOldestPostAndLatestPost(
             $cacheKey,
             $ttl,
             function () use ($blog, $nowDate, $page, $perpage) {
+                $beginningOfTime = Chronos::create(1970, 1, 1);
```
Comment (votemike) [inline src/BlogsService/Service/PostService.php]: 😸

File: src/BlogsService/Repository/AbstractRepository.php (hunk 6 above)
Comment (mattgreenham) [reply src/BlogsService/Repository/AbstractRepository.php]: Changed as discussed IRL

File: src/BlogsService/Service/PostService.php (hunk 7 above)
Comment (mattgreenham) [reply src/BlogsService/Service/PostService.php]: Yes, as we spoke about IRL

File: src/BlogsService/Repository/PostRepository.php [hunk 9]
```diff
@@ -50,46 +50,54 @@ public function getPostByGuid(string $guid, string $blogId = ''): ?ResponseInter
 
     public function getPostsBetween(
         string $blogId,
-        Chronos $afterDate,
-        Chronos $beforeDate,
+        array $ranges,
```
Comment (votemike) [inline src/BlogsService/Repository/PostRepository.php]: I think this is non-obvious enough to require a doc-block

Comment (mattgreenham) [reply src/BlogsService/Repository/PostRepository.php]: Added.

**Pull request merged**